feat: add prune check to CI

jknphy opened the pull request:
This PR adds a `ts-prune` check to the CI to detect unused code.

jknphy commented on pom/product-selection-page.ts:
```diff
@@ -0,0 +1,39 @@
+
+export class ProductSelectionPage {
+    readonly page: Page;
+    readonly choseLeap: Locator<Element>
```
`leapText` should be enough, if it is not a selector or some html element that we can name easily, let's go for text.
another thing is avoid capitilizing words, it is messy specially with the camel style, leap, microos is better.

jknphy commented on pom/login-as-root-page.ts:
```diff
@@ -0,0 +1,23 @@
+import puppeteer, { Puppeteer, type Browser, type Locator, type Page, type Product } from "puppeteer-core";
```
Puppeteer, Browser and Product are not used.

jknphy commented on pom/product-selection-page.ts:
```diff
@@ -0,0 +1,39 @@
+        await this.choseTumbleweed.click();
+        await this.selectButton.click();
+    }
+
```
blank line.

jknphy commented on src/pom-install.ts:
```diff
@@ -0,0 +1,151 @@
+import { LoginAsRootPage } from "../pom/login-as-root-page.js";
```
to properly review this class, you need to edit `default_installation.ts` instead of creating new file.

jknphy commented on dist/defult_installation.cjs:
this file seems a subproduct of some mistyping. Please drop it.

jknphy commented on src/default_installation.ts:
```diff
@@ -1,6 +1,7 @@
-// FIXME: for dumping the data when the test fails
-// import fs from "fs";
-// import path from "path";
+import { LoginAsRootPage } from "../pages/login-as-root-page";
+import { ProductSelectionPage } from "../pages/product-selection-page";
```
please add these two imports as last ones after `assert` so we order the import from more general ones to most specialized of our application.

jknphy commented on src/default_installation.ts:
```diff
@@ -67,20 +68,19 @@ function browserSettings(name: string): BrowserSettings {
 const agamaPassword = process.env.AGAMA_PASSWORD || "nots3cr3t";
 const agamaBrowser = process.env.AGAMA_BROWSER || "firefox";
-const agamaSlowMo = parseInt(process.env.AGAMA_SLOWMO || "0");
+const agamaSlowMo = parseInt(process.env.AGAMA_SLOWMO || '0');
```
let's drop this change, double quotes is fine.

jknphy commented on src/default_installation.ts:
```diff
@@ -67,20 +68,19 @@ function browserSettings(name: string): BrowserSettings {
+const agamaUserPassword = "nots3cr3t";
 
 describe("Agama test", function () {
+
```
no need for starting the method with one empty line (we will have linter and code style in some ci in the future I hope :) )

jknphy commented on src/default_installation.ts:
```diff
@@ -1,6 +1,7 @@
-// import path from "path";
+import { LoginAsRootPage } from "../pages/login-as-root-page";
+import { ProductSelectionPage } from "../pages/product-selection-page";
+import fs from "fs";
```
for this file I'm under the impression that you didn't rebase, it is better if you copy/paste the latest version and put on top your changes one by one.

jknphy commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 log/
 node_modules/
+*.cjs
```
add it back when you are completed the rebase.

jknphy commented on pages/overiew-page.ts:
```diff
@@ -0,0 +1,18 @@
+
+    constructor(page: Page) {
+        this.page = page;
+        this.installText = page.locator("button::-p-text('Install')")
```
this is a button, we should follow somehow what is in the screen, sometimes is just text unless better name, but in this case `installButton`.

jknphy commented on pages/overiew-page.ts:
```diff
@@ -0,0 +1,18 @@
+import puppeteer, { type Locator, type Page } from "puppeteer-core";
+
+// Note that this is just a minimal version to be used for a
```
all our POM are like that, we don't need to implement fully the whole screen, could you please remove this two comments?

jknphy commented on pages/overiew-page.ts:
```diff
@@ -0,0 +1,18 @@
+    async clickOnInstall() {
+        await this.installText.click();
+    }
+}
```
could you please configure your edit to leave an extra blank line after saving it?

jknphy commented on pages/overiew-page.ts:
```diff
@@ -0,0 +1,18 @@
+        this.installText = page.locator("button::-p-text('Install')")
+    }
+
+    async clickOnInstall() {
```
we need functional names, this would be just `install`.

jknphy commented on pages/overview-navigation.ts:
```diff
@@ -0,0 +1,24 @@
+// Note that this is just a minimal version to be used for a
+// default_installation.
+
+export class OverviewNavigation {
```
SidebarPage, sidebar from https://yast.opensuse.org/blog/2024-06-28/agama-9 and all the reusable blocks in POM are called pages, we should not make differenciation if it is a dialog or anything else, we messed up big hehe with our Perl POM regarding this.

jknphy commented on pages/overview-navigation.ts:
```diff
@@ -0,0 +1,24 @@
+        this.usersLink = page.locator("a[href='#/users']");
+    }
+
+    async clickOnOverview() {
```
`navigateToOverview` & `navigateToUsers`

jknphy commented on pages/users-page.ts:
```diff
@@ -0,0 +1,22 @@
+    constructor(page: Page) {
+        this.page = page;
+        this.firstUserLink = page.locator("a[href='#/users/first']");
+        this.setPasswordText = page.locator("button::-p-text(Set a password)");
```
as for now we don't have composition with two pages here, better to differenciate with:
`setAPasswordButton` (notice the 'a' to follow what is in the screen).

jknphy commented on pages/users-page.ts:
```diff
@@ -0,0 +1,22 @@
+
+    constructor(page: Page) {
+        this.page = page;
+        this.firstUserLink = page.locator("a[href='#/users/first']");
```
this looks to me a button ...
should be something like `defineAuserNowButton`, could you please check?

rakoenig replied:
I looks like a button, but it isn't. See https://progress.opensuse.org/issues/163946#note-15

jknphy commented on pages/users-page.ts:
```diff
@@ -0,0 +1,22 @@
+        this.page = page;
+        this.firstUserLink = page.locator("a[href='#/users/first']");
+        this.setPasswordText = page.locator("button::-p-text(Set a password)");
+        this.uploadSshKeyText = page.locator("button::-p-text(Upload a SSH Public Key))");
```
we don't use this, I don't recommend create locator that we don't use, the next person might find them inconvenient and not really aware if it is used somewhere else...but not big deal. We could drop it though for simplicity.

jknphy commented on pages/users-page.ts:
```diff
@@ -0,0 +1,22 @@
+        this.uploadSshKeyText = page.locator("button::-p-text(Upload a SSH Public Key))");
+    }
+
+    async goRootPasswordDialog() {
```
`setPassword`

jknphy commented on src/default_installation.ts:
```diff
@@ -105,8 +113,11 @@ const agamaInstall = booleanEnv("AGAMA_INSTALL", true);
 
 const agamaUser = "bernhard";
 const agamaUserFullName = "Bernhard M. Wiedemann";
+const agamaUserPassword = "nots3cr3t";
```
we always use the same password for user than for root, so I dropped this on purpose, in the future when we have test specific for users we can consider it, at the moment is not needed, openQA also use the same for both by default.

jknphy commented on src/default_installation.ts:
```diff
@@ -105,8 +113,11 @@ const agamaInstall = booleanEnv("AGAMA_INSTALL", true);
 
 describe("Agama test", function () {
+
+
```
extra blank lines unneeded.

jknphy commented on src/default_installation.ts:
```diff
@@ -116,7 +127,7 @@ describe("Agama test", function () {
       slowMo: agamaSlowMo,
       defaultViewport: {
-        width: 1280,
+        width: 1200,
```
I think the value is fine, no need to change, developers set it in that way.

jknphy commented on src/default_installation.ts:
```diff
@@ -127,18 +138,19 @@ describe("Agama test", function () {
   after(async function () {
-    await page.close();
-    await browser.close();
+    // await page.close();
+    // await browser.close();
```
please remove the comments for this code.

jknphy commented on src/default_installation.ts:
```diff
@@ -171,78 +183,29 @@ describe("Agama test", function () {
-    await page.locator("input#password").fill(agamaPassword);
-    await page.locator("input#passwordConfirmation").fill(agamaPassword);
-    await page.locator("button::-p-text(Confirm)").click();
+  it("should display user dialog", async function () {
```
from this point you don't need to change the test...only to inject the POM, let's not change the flow.

jknphy commented on src/default_installation.ts:
```diff
@@ -171,78 +180,54 @@ describe("Agama test", function () {
+  it("should allow setting the root password", async function () {
+    let userSection = new UsersPage(page);
+    await userSection.setPassword();
+    let passwordPopup = new RootPasswordDialog(page);
```
we will think in the future about some strategy to avoid cluttering the test with class instantiations, could you please add always the `new`  lines at the beginning, so the test is more readable in that way.

jknphy commented on src/pages/root-password-page.ts:
```diff
@@ -0,0 +1,23 @@
+import puppeteer, { type Locator, type Page } from "puppeteer-core";
+
+export class RootPasswordDialog {
```
s/RootPasswordDialog/RootPasswordPage/
we always use page as mentioned before, not any other variant, in POM a page is a whole page or a part of a page.

jknphy commented on src/pages/sidebar-page.ts:
```diff
@@ -0,0 +1,21 @@
+import puppeteer, { type Locator, type Page } from "puppeteer-core";
+
+export class Sidebar {
```
s/Sidebar/SidebarPage/

jknphy commented on src/default_installation.ts:
```diff
@@ -171,78 +180,54 @@ describe("Agama test", function () {
-    await page.locator("button::-p-text(Confirm)").click();
+  it("should display user dialog", async function () {
+    let sidebar = new Sidebar(page);
+    await sidebar.navigateToUsers();
```
we need to add back this code to be able to debug better.

jknphy commented:
I would suggest here to move forward that you close this PR and send small mergeable increments.

Otherwise you will keep rebasing code and get stuck with new things, the project is in a state that is changing too much.
It will be more effective to send one PR for each screen that you convert in in a page object, you can just go literally in the order that is developed the script and pick single one, for example with those promises you might just try to access with the page to one page at a time. There are plenty of more task to keep going. Do you agree with this strategy?

rakoenig commented:
Agreed. Let me close this and iterate through the missing steps.

okynos committed:
Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Adapt all interactive installations to rearranged storage section (test_extended)

https://progress.opensuse.org/issues/189792

Squashed commit of the following:

commit 458cd13bfecf3634b698bb68f23b6a78c3a3ec09
Author: José Luis Fernández Aguilera <jose.fernandez1@suse.com>
Date:   Mon Nov 3 09:13:16 2025 +0100

    Added requested changes

commit 280c6961a464513e9e14fe9a7750888d2ae2fe25
Author: José Luis Fernández Aguilera <jose.fernandez1@suse.com>
Date:   Thu Oct 30 12:00:29 2025 +0100

    Adapt all interactive installations to rearranged storage section (test_extended)

    https://progress.opensuse.org/issues/189792

    Adapt all interactive installations to rearranged storage section (test_extended)

    https://progress.opensuse.org/issues/189792

    Squashed commit of the following:

    commit 4add1df7bc18d65e9696d0a8746dd7ffce2653a1
    Author: José Luis Fernández Aguilera <jose.fernandez1@suse.com>
    Date:   Tue Oct 21 12:56:31 2025 +0200

        Adapt all interactive installations to rearranged storage section

        https://progress.opensuse.org/issues/189792

        Adapt all interactive installations to rearranged storage section

        https://progress.opensuse.org/issues/189792

        Adapt all interactive installations to rearranged storage section

        https://progress.opensuse.org/issues/189792

        Squashed commit of the following:

        commit 0f9cd9e863e5533d6bead20d9b9e8659ae5e6f7f
        Author: José Luis Fernández Aguilera <jose.fernandez1@suse.com>
        Date:   Tue Oct 14 16:54:13 2025 +0200

            Adapt all interactive installations to rearranged storage section

            https://progress.opensuse.org/issues/189792

            Adapt all interactive installations to rearranged storage section

            https://progress.opensuse.org/issues/189792

            Adapt all interactive installations to rearranged storage section

            https://progress.opensuse.org/issues/189792

            Added requested changes

            Applied requested comments

            Fixed span selector

        Removed unused function

        Adapt all interactive installations to rearranged storage section

        https://progress.opensuse.org/issues/189792

        Squashed commit of the following:

        commit c9f6f36310f4283852f8e84ffd8029b0abae651d
        Author: José Luis Fernández Aguilera <jose.fernandez1@suse.com>
        Date:   Tue Oct 14 16:54:13 2025 +0200

            Adapt all interactive installations to rearranged storage section

            https://progress.opensuse.org/issues/189792

            Adapt all interactive installations to rearranged storage section

            https://progress.opensuse.org/issues/189792

            Added initial modifications

            Removed icon in text match

    Fix linter

    Adapt all interactive installations to rearranged storage section (test_extended)

    https://progress.opensuse.org/issues/189792

    Fix encryption check

    Fix destructive actions

    Changed destructive list selector

    Added requested changes

    Applied requested comment

Updated JS files

Applied requested comments

Added branching of prod and new UI

Added product version

Fix variable name

Changed naming

Changed name

jknphy committed:
Merge pull request #175 from okynos/189792-adapt-code-versions

Adapt all interactive installations to rearranged storage section (test_extended)

jknphy committed:
refactor: Implement Strategy pattern for version handling

Implement the Strategy and Factory patterns to handle different test logic for different product versions.

- Add IProductTestStrategy interface.
- Add PreReleaseStrategy and StableReleaseStrategy concrete classes.
- Add ProductStrategyFactory to create the appropriate strategy.
- Refactor test_extended.ts to use the factory and strategy.

jknphy committed:
Merge pull request #176 from jknphy/version-handling

refactor: Implement Strategy pattern for version handling

jknphy committed:
feat: add prune check to CI
